File: DangerSnake/DangerSnake/Core/GridPos.swift
import Foundation

struct GridPos: Hashable, Equatable, Sendable {
    var x: Int
    var y: Int

    static func + (lhs: GridPos, rhs: GridPos) -> GridPos {
        GridPos(x: lhs.x + rhs.x, y: lhs.y + rhs.y)
    }

    static func - (lhs: GridPos, rhs: GridPos) -> GridPos {
        GridPos(x: lhs.x - rhs.x, y: lhs.y - rhs.y)
    }

    func manhattan(_ other: GridPos) -> Int {
        abs(x - other.x) + abs(y - other.y)
    }
}

enum Directions {
    static let up = GridPos(x: 0, y: 1)
    static let down = GridPos(x: 0, y: -1)
    static let left = GridPos(x: -1, y: 0)
    static let right = GridPos(x: 1, y: 0)
    static let all = [up, down, left, right]

    static func opposite(_ dir: GridPos) -> GridPos {
        if dir == up { return down }
        if dir == down { return up }
        if dir == left { return right }
        if dir == right { return left }
        return dir
    }
}

File: Unity/Assets/Scripts/Core/GridPos.cs
using UnityEngine;

namespace DangerSnake.Core
{
    public readonly struct GridPos
    {
        public readonly int X;
        public readonly int Y;

        public GridPos(int x, int y)
        {
            X = x;
            Y = y;
        }

        public static GridPos operator +(GridPos a, GridPos b) => new GridPos(a.X + b.X, a.Y + b.Y);
        public static GridPos operator -(GridPos a, GridPos b) => new GridPos(a.X - b.X, a.Y - b.Y);
        public static bool operator ==(GridPos a, GridPos b) => a.X == b.X && a.Y == b.Y;
        public static bool operator !=(GridPos a, GridPos b) => !(a == b);

        public int Manhattan(GridPos other) => Mathf.Abs(X - other.X) + Mathf.Abs(Y - other.Y);

        public override bool Equals(object obj) => obj is GridPos other && this == other;
        public override int GetHashCode() => (X * 397) ^ Y;
        public override string ToString() => $"({X},{Y})";
    }

    public static class Directions
    {
        public static readonly GridPos Up = new GridPos(0, 1);
        public static readonly GridPos Down = new GridPos(0, -1);
        public static readonly GridPos Left = new GridPos(-1, 0);
        public static readonly GridPos Right = new GridPos(1, 0);

        public static readonly GridPos[] All = { Up, Down, Left, Right };

        public static GridPos Opposite(GridPos dir)
        {
            if (dir == Up) return Down;
            if (dir == Down) return Up;
            if (dir == Left) return Right;
            if (dir == Right) return Left;
            return dir;
        }
    }
}
